lib/datasource/artifactory/index.ts

```diff
@@ -0,0 +1,97 @@
+import { logger } from '../../logger';
+import { ExternalHostError } from '../../types/errors/external-host-error';
+import { cache } from '../../util/cache/package/decorator';
+import { parse } from '../../util/html';
+import { HttpError } from '../../util/http/types';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
+import { datasource, defaultRegistryUrl } from './common';
+
+export class ArtifactoryDatasource extends Datasource {
+  static readonly id = datasource;
+
+  constructor() {
+    super(datasource);
+  }
+
+  override readonly customRegistrySupport = true;
+
+  override readonly defaultRegistryUrls = [defaultRegistryUrl];
+
+  override readonly caching = true;
+
+  override readonly registryStrategy = 'merge';
+
+  @cache({
+    namespace: `datasource-${datasource}`,
+    key: ({ registryUrl }: GetReleasesConfig) => `${registryUrl}`,
+  })
+  async getReleases({
+    lookupName,
+    registryUrl,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    if (registryUrl === defaultRegistryUrl) {
+      logger.warn(
+        'artifactory datasource requires custom registryUrl. Skipping datasource'
+      );
+      return null;
+    }
+
+    const url = `${registryUrl}/${lookupName}`;
+
+    const result: ReleaseResult = {
+      homepage: 'https://jfrog.com/artifactory',
+      releases: [],
+    };
+    try {
+      const response = await this.http.get(url);
+      const body = parse(ArtifactoryDatasource.cleanSimpleHtml(response.body));
+      const nodes = body.querySelectorAll('a');
+
+      let candidates: string[] = [];
+      nodes.forEach((node) => candidates.push(node.innerHTML));
+
+      // filter out hyperlink to navigate to parent folder
+      candidates = candidates.filter(
+        (candidate) => candidate !== '../' && candidate !== '..'
+      );
+
+      candidates.forEach((candidate) => {
+        const parsedCandidate: string =
+          candidate.slice(-1) === '/' ? candidate.slice(0, -1) : candidate;
+
+        const thisRelease: Release = {
+          version: '',
+        };
+        thisRelease.version = parsedCandidate;
+        result.releases.push(thisRelease);
+      });
+
+      if (result.releases.length) {
+        logger.debug('Found ' + String(result.releases.length) + ' releases');
+      } else {
+        logger.debug(
+          'Not found any version of ' + lookupName + ' under ' + url
+        );
+      }
+    } catch (err) {
+      // istanbul ignore else: not testable with nock
+      if (err instanceof HttpError) {
+        if (err.response?.statusCode !== 404) {
+          throw new ExternalHostError(err);
```

Comment:
This would mean that if a package is not found on Artifactory then Renovate will abort the entire repo each time and never update anything. I don't think we want this.

Reply:
Do you mean for the case of status code 404?

I see, if there are other managers running they won't run anything after this error. It is more convenient to return nothing and log with a message.
I don't know how much about the error handling in Renovate, I saw that the error was caught and the renovate call didn't return any error code, so I thought that after the error everything was running as if there wasn't any error. Thanks for bringing this up.

Reply:
`ExternalHostError` causes an abort of a run. I think best to instead return `null` and keep going

Reply:
Applied suggestion.